File: backend/app/schemas/publish.py
from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field, model_validator

JobStatus = Literal["queued", "posting", "posted", "failed", "skipped"]
SourceKind = Literal["single", "bulk_row"]
BulkRunStatus = Literal[
    "queued", "running", "paused", "cancelled", "done", "failed"
]
RowFilter = Literal["all", "selected", "range"]
CellFilter = Literal["all", "unpublished", "failed"]
PublishMode = Literal["single", "multi"]
# Per-run publish operation.
#   'create' — POST a new post / page.
#   'update' — For WP: resolve an existing post via lookup_kind+lookup_column_id
#              and PATCH it. For Custom CMS: send action='update' with the
#              upstream id supplied as the `id` field-to-column mapping; no
#              find_post pre-flight (the upstream resolves the id itself).
#   'upsert' — Custom CMS only. Server-side slug → existing-page lookup with
#              fallback to create when the slug isn't found. WP has no native
#              upsert and the API rejects this combination at run creation.
PublishOperation = Literal["create", "update", "upsert"]
PublishLookupKind = Literal["id", "slug"]
# How to react when a Create row's slug already exists on the target
# (in its language, per the find_post language filter). 'create' = always
# POST (WP auto-suffixes); 'skip' = log as skipped; 'update' = PATCH the
# existing post.
OnSlugConflict = Literal["create", "skip", "update"]


class BulkPublishRequest(BaseModel):
    table_id: int
    # 'single' (today's behavior) or 'multi' (domain + profile come from
    # cells on each row).
    mode: PublishMode = "single"

    # Single-mode targets — required when mode='single', ignored otherwise.
    domain_id: int | None = None
    profile_name: str | None = None  # WP only; '' or omitted for Custom

    # Multi-mode column refs — required when mode='multi', ignored otherwise.
    domain_column_id: int | None = None
    profile_column_id: int | None = None
    # Per-row language column (multi mode only). Cell value is lowercased
    # + trimmed and matched against the resolved domain's languages.
    # Run-level `language` becomes the fallback when this column is unset
    # OR not in multi mode.
    language_column_id: int | None = None

    language: str | None = None

    row_filter: RowFilter = "all"
    selection: dict[str, Any] | None = None  # {row_ids:[...]} or {start,end}
    cell_filter: CellFilter = "all"

    field_to_column: dict[str, int] = Field(default_factory=dict)
    back_fill: dict[str, int] = Field(default_factory=dict)

    save_mapping: bool = True

    # Create vs Update. Update mode resolves each row to an existing WP post
    # via (lookup_kind, lookup_column_id) and PATCHes it instead of posting
    # a new one. WP-only — Custom CMS in update mode is rejected at run
    # creation by the API layer (it has no PATCH/find convention).
    operation: PublishOperation = "create"
    lookup_kind: PublishLookupKind | None = None
    lookup_column_id: int | None = None

    # Slug-conflict handling on Create. Pre-checks each row's slug via the
    # WP REST API and either skips a duplicate or PATCHes it. Adds one GET
    # per row, language-aware via find_post.
    on_slug_conflict: OnSlugConflict = "create"

    @model_validator(mode="after")
    def _validate_mode(self) -> "BulkPublishRequest":
        if self.mode == "single":
            if self.domain_id is None:
                raise ValueError("domain_id is required in single mode")
            # multi-only fields ignored even if sent
        else:
            if self.domain_column_id is None:
                raise ValueError("domain_column_id is required in multi mode")
            # profile_column_id required for WP; we can't validate per-CMS here
            # (need the resolved domain), so the API validates after looking
            # up domains. profile_column_id presence is checked there.

        # lookup_kind / lookup_column_id are WP-specific (find_post by id|slug).
        # Custom CMS update sends the upstream id via the `id` field-to-column
        # mapping instead — no find_post needed. We can't see cms_type at this
        # layer, so we don't enforce lookup_* presence here; the API layer
        # validates it for WP runs and the service-layer Custom-CMS branch
        # ignores those fields entirely.
        if self.on_slug_conflict != "create":
            # Skip / Update behaviors only apply to Create-mode runs. In
            # Update mode every row already targets an existing post; a
            # second pre-check is redundant and likely a UI mistake.
            if self.operation != "create":
                raise ValueError(
                    "on_slug_conflict applies only when operation='create'. "
                    "Update-mode runs already target an existing post."
                )
            # Without `slug` in field_to_column there's nothing to pre-check
            # (WP would derive the slug from the title server-side). Fail
            # explicitly so the user knows the toggle has no effect.
            if "slug" not in self.field_to_column:
                raise ValueError(
                    "on_slug_conflict requires 'slug' to be mapped in "
                    "field_to_column. Map a column to the slug field, or "
                    "set on_slug_conflict='create'."
                )
        return self


class BulkRunSummary(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    created_at: datetime
    started_at: datetime | None
    finished_at: datetime | None
    table_id: int
    mode: PublishMode = "single"
    domain_id: int | None
    domain_name: str | None = None
    table_name: str | None = None
    profile_name: str | None
    language: str | None
    status: BulkRunStatus
    total: int
    done: int
    failed: int
    skipped: int
    error: str | None
    created_by_id: int | None
    operation: PublishOperation = "create"
    lookup_kind: PublishLookupKind | None = None
    lookup_column_id: int | None = None
    language_column_id: int | None = None
    on_slug_conflict: OnSlugConflict = "create"


class ByDomainStat(BaseModel):
    """Per-domain breakdown for a multi-mode run's detail view."""

    domain_id: int | None
    domain_name: str | None  # null when domain_id is null (unresolved)
    total: int
    posted: int
    failed: int


class BulkRunDetail(BulkRunSummary):
    row_filter: RowFilter
    selection: dict[str, Any] | None
    cell_filter: CellFilter
    field_to_column: dict[str, int]
    back_fill: dict[str, int]

    domain_column_id: int | None = None
    profile_column_id: int | None = None
    # Empty list for single-mode runs (UI hides the panel anyway).
    by_domain: list[ByDomainStat] = Field(default_factory=list)


class BulkRunListResponse(BaseModel):
    items: list[BulkRunSummary]
    total: int
    page: int
    page_size: int


class PublishMapping(BaseModel):
    field_to_column: dict[str, int] = Field(default_factory=dict)
    back_fill: dict[str, int] = Field(default_factory=dict)
    language: str | None = None
    # Multi-mode only: which columns held domain / profile last time. Null
    # for single-mode mappings; the UI ignores them in that case.
    domain_column_id: int | None = None
    profile_column_id: int | None = None
    language_column_id: int | None = None
    # Remembered Create/Update choice + lookup target, so the modal restores
    # the same shape next time the user opens it for this (table, mode).
    operation: PublishOperation = "create"
    lookup_kind: PublishLookupKind | None = None
    lookup_column_id: int | None = None
    on_slug_conflict: OnSlugConflict = "create"


class PublishDefaults(BaseModel):
    """Global publish rate-limit defaults. Mirrored on each domain as a nullable
    override; NULL on the domain means "use the value from here".
    """

    requests_per_minute: int = Field(..., ge=1, le=100000)
    max_concurrency: int = Field(..., ge=1, le=1000)
    inter_request_delay_ms: int = Field(..., ge=0, le=600000)
    retry_max_attempts: int = Field(..., ge=0, le=20)
    backoff_base_ms: int = Field(..., ge=0, le=600000)
    backoff_jitter_ms: int = Field(..., ge=0, le=600000)
    respect_retry_after: bool


class PublishSingleRequest(BaseModel):
    domain_id: int
    fields: dict[str, Any] = Field(default_factory=dict)
    language: str | None = None
    profile_name: str | None = None  # WP only; Custom ignores
    source_ref: dict[str, Any] | None = None  # e.g. {"generation_id": 5, "prompt_id": 3}


class PublishJobRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    created_at: datetime
    finished_at: datetime | None
    domain_id: int | None
    domain_name: str | None = None
    source_kind: SourceKind
    source_ref: dict[str, Any] | None
    status: JobStatus
    language: str | None
    cms_post_id: str | None
    cms_post_url: str | None
    error: str | None
    warnings: list[str] | None = None
    profile_name: str | None = None
    created_by_id: int | None


class PublishJobDetail(PublishJobRead):
    payload_sent: dict[str, Any] | None
    response_json: dict[str, Any] | None


class PublishJobListResponse(BaseModel):
    items: list[PublishJobRead]
    total: int
    page: int
    page_size: int

File: frontend/lib/publish.ts
import { api } from "@/lib/api";

export type JobStatus = "queued" | "posting" | "posted" | "failed" | "skipped";
export type SourceKind = "single" | "bulk_cell";

export interface PublishJob {
  id: number;
  created_at: string;
  finished_at: string | null;
  domain_id: number | null;
  domain_name: string | null;
  source_kind: SourceKind;
  source_ref: Record<string, unknown> | null;
  status: JobStatus;
  language: string | null;
  cms_post_id: string | null;
  cms_post_url: string | null;
  /**
   * Upstream HTTP status code (100–599). NULL for rows that landed
   * before backend migration 0026 (no way to backfill), or for rows
   * where the request never reached an HTTP layer (e.g. an SSRF
   * pre-flight rejection). Surfaces in the run-detail table next to
   * the status badge so a "posted 201" vs "failed 403" is visible at a
   * glance without parsing the error string.
   */
  status_code: number | null;
  error: string | null;
  warnings: string[] | null;
  profile_name: string | null;
  created_by_id: number | null;
  /** The slug actually sent to the CMS (post-normalization). null when no slug
   *  field was sent (e.g. an empty slug dropped from the body). */
  slug: string | null;
}

export interface PublishJobDetail extends PublishJob {
  payload_sent: Record<string, unknown> | null;
  response_json: Record<string, unknown> | null;
  /** Copy-pasteable curl reproducing the exact request sent for this row
   *  (method + URL + headers with auth masked + the real JSON body). null
   *  when nothing was sent yet or the target can't be resolved. */
  curl_preview: string | null;
}

export interface PublishJobListResponse {
  items: PublishJob[];
  total: number;
  page: number;
  page_size: number;
}

export interface PublishSinglePayload {
  domain_id: number;
  fields: Record<string, unknown>;
  language?: string | null;
  profile_name?: string | null;
  source_ref?: Record<string, unknown> | null;
}

export function publishSingle(payload: PublishSinglePayload) {
  return api<PublishJobDetail>("/publish/single", { method: "POST", body: payload });
}

export function listPublishJobs(opts: {
  page?: number;
  page_size?: number;
  status?: JobStatus;
  source_kind?: "single" | "bulk_row" | "bulk_cell";
  domain_id?: number;
  run_id?: number;
  generation_id?: number;
} = {}) {
  const p = new URLSearchParams();
  for (const [k, v] of Object.entries(opts)) {
    if (v === undefined || v === null) continue;
    const s = String(v);
    if (!s) continue;
    p.set(k, s);
  }
  const qs = p.toString();
  return api<PublishJobListResponse>(`/publish/jobs${qs ? `?${qs}` : ""}`);
}

export function getPublishJob(id: number) {
  return api<PublishJobDetail>(`/publish/jobs/${id}`);
}

export function deletePublishJob(id: number) {
  return api<void>(`/publish/jobs/${id}`, { method: "DELETE" });
}

export function clearCompletedPublishJobs(opts: {
  source_kind?: "single" | "bulk_row" | "bulk_cell";
} = {}) {
  const qs = opts.source_kind ? `?source_kind=${opts.source_kind}` : "";
  return api<{ deleted: number }>(`/publish/jobs/completed${qs}`, {
    method: "DELETE",
  });
}

export interface PublishDefaults {
  requests_per_minute: number;
  max_concurrency: number;
  inter_request_delay_ms: number;
  retry_max_attempts: number;
  backoff_base_ms: number;
  backoff_jitter_ms: number;
  respect_retry_after: boolean;
}

export function getPublishDefaults() {
  return api<PublishDefaults>("/publish/defaults");
}

export function setPublishDefaults(payload: PublishDefaults) {
  return api<PublishDefaults>("/publish/defaults", {
    method: "PUT",
    body: payload,
  });
}
